include/swoole_socket.h

```diff
@@ -161,107 +166,122 @@ struct Socket {
-    case ECONNRESET:
+
+    static inline int is_dgram(uint8_t type) {
+        return (type == SW_SOCK_UDP || type == SW_SOCK_UDP6 || type == SW_SOCK_UNIX_DGRAM);
+    }
+
+    static inline int is_stream(uint8_t type) {
+        return (type == SW_SOCK_TCP || type == SW_SOCK_TCP6 || type == SW_SOCK_UNIX_STREAM);
+    }
+
+    ssize_t sendto_blocking(const void *buf, size_t n, int flag, struct sockaddr *addr, socklen_t addr_len);
+
+    inline ssize_t sendto(const char *dst_host, int dst_port, const char *data, uint32_t len) {
+        Address addr;
+        if (!addr.assign(socket_type, dst_host, dst_port)) {
+            return SW_ERR;
+        }
+        return sendto(&addr, data, len);
+    }
+
+    inline ssize_t sendto(Address *dst_addr, const char *data, uint32_t len) {
+        return ::sendto(fd, data, len, 0, (struct sockaddr *) &dst_addr->addr, dst_addr->len);
+    }
+
+    inline int catch_error(int err) {
+        switch (err) {
+        case EFAULT:
+            abort();
+            return SW_ERROR;
+        case EBADF:
+        case ECONNRESET:
 #ifdef __CYGWIN__
-    case ECONNABORTED:
+        case ECONNABORTED:
 #endif
-    case EPIPE:
-    case ENOTCONN:
-    case ETIMEDOUT:
-    case ECONNREFUSED:
-    case ENETDOWN:
-    case ENETUNREACH:
-    case EHOSTDOWN:
-    case EHOSTUNREACH:
-    case SW_ERROR_SSL_BAD_CLIENT:
-    case SW_ERROR_SSL_RESET:
-        return SW_CLOSE;
-    case EAGAIN:
+        case EPIPE:
+        case ENOTCONN:
+        case ETIMEDOUT:
+        case ECONNREFUSED:
+        case ENETDOWN:
+        case ENETUNREACH:
+        case EHOSTDOWN:
+        case EHOSTUNREACH:
+        case SW_ERROR_SSL_BAD_CLIENT:
+        case SW_ERROR_SSL_RESET:
+            return SW_CLOSE;
+        case EAGAIN:
 #ifdef HAVE_KQUEUE
-    case ENOBUFS:
+        case ENOBUFS:
 #endif
-    case 0:
-        return SW_WAIT;
-    default:
-        return SW_ERROR;
-    }
-}
-
-static sw_inline int swSocket_get_domain_and_type(enum swSocket_type type, int *sock_domain, int *sock_type) {
-    switch (type) {
-    case SW_SOCK_TCP6:
-        *sock_domain = AF_INET6;
-        *sock_type = SOCK_STREAM;
-        break;
-    case SW_SOCK_UNIX_STREAM:
-        *sock_domain = AF_UNIX;
-        *sock_type = SOCK_STREAM;
-        break;
-    case SW_SOCK_UDP:
-        *sock_domain = AF_INET;
-        *sock_type = SOCK_DGRAM;
-        break;
-    case SW_SOCK_UDP6:
-        *sock_domain = AF_INET6;
-        *sock_type = SOCK_DGRAM;
-        break;
-    case SW_SOCK_UNIX_DGRAM:
-        *sock_domain = AF_UNIX;
-        *sock_type = SOCK_DGRAM;
-        break;
-    case SW_SOCK_TCP:
-        *sock_domain = AF_INET;
-        *sock_type = SOCK_STREAM;
-        break;
-    default:
-        return SW_ERR;
+        case 0:
+            return SW_WAIT;
+        default:
+            return SW_ERROR;
+        }
     }
 
-    return SW_OK;
-}
+    static inline int get_domain_and_type(enum swSocket_type type, int *sock_domain, int *sock_type) {
```

Comment:
Perhaps passing references is more expressive

Reply:
Follow the google specification, if it is a write operation to pass pointer, read-only operation to pass reference